### ISSUE TITLE
csv: Inconsistency re QUOTE_NONNUMERIC

### ISSUE THREAD

**Issue by 4c93acd5-d8bc-4c24-9607-569c85e01bc1** [python/cpython#74232]
BPO | [30046](https://bugs.python.org/issue30046)
--- | :---
Nosy | @serhiy-storchaka, @zhangyangyu, @tlotze, @corona10

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-04-11.20:41:55.773>
labels = ['type-bug', 'library']
title = 'csv: Inconsistency re QUOTE_NONNUMERIC'
updated_at = <Date 2017-04-19.07:11:22.021>
user = 'https://github.com/tlotze'
```

bugs.python.org fields:
```python
activity = <Date 2017-04-19.07:11:22.021>
actor = 'corona10'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-04-11.20:41:55.773>
creator = 'tlotze'
dependencies = []
files = []
hgrepos = []
issue_num = 30046
keywords = []
message_count = 5.0
messages = ['291516', '291590', '291862', '291864', '291865']
nosy_count = 4.0
nosy_names = ['serhiy.storchaka', 'xiang.zhang', 'tlotze', 'corona10']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue30046'
versions = []
```

</p></details>

### Linked PRs
* gh-134422
* gh-134963
* gh-135048
* gh-135049

**Comment by 4c93acd5-d8bc-4c24-9607-569c85e01bc1** [comment]
A csv.writer with quoting=csv.QUOTE_NONNUMERIC does not quote boolean values, which makes a csv.reader with the same quoting behaviour fail on that value:

-------- csv.py \----------

```py
import csv
import io


f = io.StringIO()

writer = csv.writer(f, quoting=csv.QUOTE_NONNUMERIC)
writer.writerow(['asdf', 1, True])

f.seek(0)
reader = csv.reader(f, quoting=csv.QUOTE_NONNUMERIC)
for row in reader:
    print(row)
```


----------------------

```pytb
$ python3 csvbug.py 
Traceback (most recent call last):
  File "csvbug.py", line 12, in <module>
    for row in reader:
ValueError: could not convert string to float: 'True'
```


----------------------

I'd consider this inconsistency a bug, but in any case something that needs documenting.

**Comment by zhangyangyu** [comment]
boolean is not quoted since in Python it's a subclass of int so True and False are numeric. This is also the case with numeric objects defining \_\_int__ or \_\_float__ but doesn't get a corresponding string representation.

Since QUOTE_NONNUMERIC will converts data to float when reading, I think we may force the converting even when writing so the inconsistency would disappear. Or document this limitation.

**Comment by serhiy-storchaka** [comment]
This issue is not easy, it needs a thoughtful design before starting coding. I agree with Xiang's analysis and proposed solutions. But if just convert numbers to float we can get an overflow for large integers or lost precision in case of Decimal. The consumer of the CSV file may be not Python and it may support larger precision than Python float.

And it is not clear what would be better solution for enums. Should they be serialized by name or by value?

**Comment by corona10** [comment]
I would like to solve this issue. 
Is there any other way than casting specifically for the bool object?
(e.g For general ways?)

**Comment by corona10** [comment]
Oh, I read the Serhiy Storchaka 's comment just right now.

**Comment by aterrel** [comment]
I don't think it is reasonable to change the type system and this isn't a type bug perse, more of a difference in type expectations. 

I've added the note to the docs but this is my first doc fix in CPython so not sure if I did it correctly.

**Comment by serhiy-storchaka** [comment]
gh-155097 (PR #155099) makes this round-trip possible without changing the default: `bool` is still a number for quoting purposes, so `True` is still written unquoted.

A `formatter` writes bools in the form the consumer expects, and a `converter` reads them back:

```python
BOOLS = {'true': True, 'false': False}
csv.writer(f, quoting=csv.QUOTE_NONNUMERIC,
           formatter=lambda i, v: 'true' if v is True else 'false' if v is False else str(v))
csv.reader(f, quoting=csv.QUOTE_NONNUMERIC,
           converter=lambda i, s: BOOLS[s] if s in BOOLS else float(s))
```

The same applies to `Fraction`, `IntEnum` and the other types in the note added by GH-134963.
